### PR TITLE
[LibOS] Print warning when running as root user

### PR DESCRIPTION
**Pull request by dimakuv**
## Description of the changes 

It is considered a good practice to run as non-root user whenever possible. E.g., Docker recommends it in its "Best practices":
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user

Kudos to @marcelamelara for pointing this out.

## How to test this PR? 

Manually.

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)

### CODE REVIEW

**Review by oshogbo — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158#-:-NNLyiaU2Ig2qPleDh9j:bnfp4nl)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv)

___
*[`libos/src/bookkeep/libos_thread.c` line 184 at r1](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNLyZPl6JfpgKwvc_N4:-NNLyZPl6JfpgKwvc_N5:b-6utxag) ([raw file](https://github.com/gramineproject/gramine/blob/77dff3187f55ac783a1ebf11419c5bbc971d8d48/libos/src/bookkeep/libos_thread.c#L184)):*
> ```C
>         log_warning("Application starts as root (UID and/or GID are equal to 0). Running as root "
>                     "brings some risks. Consider specifying a non-root user in the manifest (use "
>                     "'loader.uid' and 'loader.gid' options).");
> ```

I wonder if we should have `log_advaice`, which users can disable in the manifest?

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NNYdm6T-gWjS-H8QmY3:beaniho) (no related file):*
Maybe just merge https://github.com/gramineproject/graphene/pull/2654 instead? Seems we forgot about it, because it got stuck on some issues with Docker, but I don't remember why the final version of it wasn't ok. @boryspoplawski do you remember maybe?

___
___
*[`libos/src/bookkeep/libos_thread.c` line 181 at r1](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYe9PV-pvbrLA6d3FR:-NNYe9PV-pvbrLA6d3FS:bmk9yxx) ([raw file](https://github.com/gramineproject/gramine/blob/77dff3187f55ac783a1ebf11419c5bbc971d8d48/libos/src/bookkeep/libos_thread.c#L181)):*
> ```C
>     }
> 
>     if (uid_int64 == 0 || gid_int64 == 0) {
> ```

Wait, but that's the user id _inside_ Gramine, isn't it?

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv and @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NNYkw1FEG2luzY87mC8:b-qbp940) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Maybe just merge https://github.com/gramineproject/graphene/pull/2654 instead? Seems we forgot about it, because it got stuck on some issues with Docker, but I don't remember why the final version of it wasn't ok. @boryspoplawski do you remember maybe?
</blockquote></details>

It wasn't merged because at that time default SGX driver you could get would be only usable as root, unless you manually configured permissions. We didn't have any guide in our docs on how to do that, so we decided to wait.
Also I admit it was rather a rage PR, after seeing a one too many report where someone run stuff as root and it changes how things works and reproducibility of an issue. After more considerations disallowing running stuff as root is IMO only annoying (especially when someone knowing what they do, really wants to run as root).

BTW, this PR makes no sense to me - it checks in Gramine uids.

___

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NN_oN-eCt8_cx9rg9Bc:b16rufi) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

It wasn't merged because at that time default SGX driver you could get would be only usable as root, unless you manually configured permissions. We didn't have any guide in our docs on how to do that, so we decided to wait.
Also I admit it was rather a rage PR, after seeing a one too many report where someone run stuff as root and it changes how things works and reproducibility of an issue. After more considerations disallowing running stuff as root is IMO only annoying (especially when someone knowing what they do, really wants to run as root).

BTW, this PR makes no sense to me - it checks in Gramine uids.
</blockquote></details>

@mkow @boryspoplawski You misunderstood the rationale for this PR.

It's not about running Gramine as host root user.

It's about running the application inside Gramine thinking it is a root user.

Note that this PR is debatable -- one can always argue that the Gramine manifest writer must do her due diligence and verify that `loader.uid` and `loader.gid` can stay as defaults (= 0, aka root user), and nothing bad/weird happens to the application.

For example, this is the case for Memcached -- it **must** be run as non-root user inside Gramine, otherwise Memcached tries to do some prohibited thing (don't remember what exactly), and Memcached fails loudly.

To make Gramine users aware of such possible behavior, I wrote this PR. Basically, when I saw that Docker documentation mentions the same issue (root user inside Docker container), I decided that it's a popular problem, and Gramine should highlight it somehow. Docker docs: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user

___
___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/1158#-NN_pYSP7Ch1dgxpRp7_):*
Maybe this commit title is misleading. Something like `Print warning when app inside Gramine starts as root` should be better.
___
*[`libos/src/bookkeep/libos_thread.c` line 181 at r1](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYe9PV-pvbrLA6d3FR:-NN_phmK6zjDdiL9YJLK:btgi96o) ([raw file](https://github.com/gramineproject/gramine/blob/77dff3187f55ac783a1ebf11419c5bbc971d8d48/libos/src/bookkeep/libos_thread.c#L181)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Wait, but that's the user id _inside_ Gramine, isn't it?
</blockquote></details>

Yes, that's the whole point of this PR. Please see my top-level comment.
___
*[`libos/src/bookkeep/libos_thread.c` line 184 at r1](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNLyZPl6JfpgKwvc_N4:-NN_plT80psTHwnBvko1:bd81iit) ([raw file](https://github.com/gramineproject/gramine/blob/77dff3187f55ac783a1ebf11419c5bbc971d8d48/libos/src/bookkeep/libos_thread.c#L184)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

I wonder if we should have `log_advaice`, which users can disable in the manifest?
</blockquote></details>

@oshogbo Sounds like an interesting idea, but I've never encountered such thing in practice. Can you give some examples of application that do it, and how exactly they use `log_advice`?

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NNaxORv-H-g_1XA3bEa:btmd5s7) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@mkow @boryspoplawski You misunderstood the rationale for this PR.

It's not about running Gramine as host root user.

It's about running the application inside Gramine thinking it is a root user.

Note that this PR is debatable -- one can always argue that the Gramine manifest writer must do her due diligence and verify that `loader.uid` and `loader.gid` can stay as defaults (= 0, aka root user), and nothing bad/weird happens to the application.

For example, this is the case for Memcached -- it **must** be run as non-root user inside Gramine, otherwise Memcached tries to do some prohibited thing (don't remember what exactly), and Memcached fails loudly.

To make Gramine users aware of such possible behavior, I wrote this PR. Basically, when I saw that Docker documentation mentions the same issue (root user inside Docker container), I decided that it's a popular problem, and Gramine should highlight it somehow. Docker docs: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user
</blockquote></details>

But how's that a good practice? You are still running as the same user on host (whatever that user is), so it changes virtually nothing.
TLDR Exploiting an app running in Gramine will have the exact same consequences, whatever you set as uid. 

___

**Review by mkow — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski, @dimakuv, and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NNb2vff6yPmDMNfBYdY:b-ltrzte) (no related file):*
> It's about running the application inside Gramine thinking it is a root user.

Ok, then I also don't understand the purpose of this PR.

___

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1158)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1158#-NNYdm6T-gWjS-H8QmY2:-NNb4O1s9RPrWSFsxFIC:b-7uk4sk) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> It's about running the application inside Gramine thinking it is a root user.

Ok, then I also don't understand the purpose of this PR.
</blockquote></details>

Ok, two reviewers are against this warning message. Let me close this PR.

___
